### DISCUSSION TITLE
using ESP32 to drive DM860h via TXS0108e LLC, anybody have any advice, cannot get pulses thru to driver... works on Nano!

### DISCUSSION THREAD

**Discussion by andymid3D** [gin66/FastAccelStepper#291] [Q&A] [answered]
anybody have any advice on what may be happening wrong here..:
have a working sketch on nano, connected direct to DM860H, Nema 34 moves perfect... as the project is growing and the Nano can't keep up, decided to go with ESP32, however cannot get the motor to move correctly, if at all.. it would seem a problem of logic voltaje levels from what I can read with my multimeter on the DIR port of the driver, but there is more to the strangeness story...

using a AMS1117 3.3V buck DC-DC regulator to power both ESP32 and LLC (and OE pin high)... and a 5v 2A PSU that powers the regulator and the LLC as well.. have 2 buttons and a pot that indicate direction and speed in sketch.. strangest thing is, that as soon as connect only GND to PUL- & DIR- on the driver, I am seeing 2-2.2v between PUL+ & -, and also DIR + & -, when I then connect to LLC 5v side, then the OUTPUTs of the 5v LLC side are actually LOWER than the INPUTS from the ESP32 - and also the inputs are as well reduced, the DIR from 3.3v to about 3v... if the driver is not connected to LLC, ESP32 outputs are correct, and LLC outputs are correct, 5v in case of DIR being positive move, 0v when negative move commanded...

the only thing can think of is maybe the LLC is using open-drain (whatever that really is..?) and should be using push-pull (pull-ups..?), as the FAS commands are very fast, but how to tell FAS to use pull-ups on ESP32...? or is this irrelevante for outputs anyway on the dev board...?

I have made (my first) repo on github to add datasheets of components used, photos of circuit, more docs maybe of interest, I will later add also 3D model when completed, etc, and can add other suggested things here too.. 

https://github.com/andymid3D/PP-motorized-injector

I consider myself a mid-novice hardware and code user, autodidact since 15 (now 52), so am very aware that I know that I don't know more than I do, so any suggestions, guidance, tutorials, etc are welcome (although life is so busy, it would have to be directly related to learning my way thru the problem to be able to take on immediately, otherwise to the pile of course I want to brush up my knowledge about, but just too scattered to have time ATM!)

**Comment by gin66** [comment]
The repo is not accessible or non-existent.

FastAccelStepper knows only Step/Dir/Enable as digital pins with some flexibility for polarity and timing of dir/enable signals.  For step-pins polarity and timing are fixed: step on low to high transition and high time between 7 and ~2000us. DM860-H works with 5-24V inputs. Perhaps need a level shifter as (https://forum.arduino.cc/t/control-nema-34-with-esp32-and-dm860t/1111838/3) ? or use a TB6600 ?

LLC stand for what ?

**Reply by andymid3D** [reply]
First time I make a repo on GIthub, thought if was private I could let people see with the link..? anyway, made public now, FWIW ;-)

I have successfully used a Nano (5v logic) to run this simple sketch with the DM860H directly from the Nano, and works wonderfully, and with the ESP32 am now using a level shifter txs0108e, as it is supposed to be fast enough to be able to use with motor pulses in this situation (3.3v to 5v logic signals), but it seems there are 2 situations, "open-drain" which is a lot slower, and "push-pull" which is the fast one (and I have no idea how to make it use the faster way)... I have tried with different level shifter, a simpler one, with the same results

LLC = logic level converter (AFAIK?), the data sheet for the tsx0108e in the repo is called "txs0108e LLC.pdf" ;-)

TB6600 3.5A @ 42v max - I cannot go down in power (DM860 is 6A @ 80V), for the use case (plastic injection) I am already at the minimum of industrial standard injection pressure with the actual setup.. anyway this driver would seem identical in its logic operation, which as mentioned, with 5v Arduino, have no problem... 

any other examples out there of using FastAccel with an external driver via step/dir, from an ESP32 and using a level shifter/ LLC? That would be really helpful.. the ESP32 pinout says that 6mA is nominal output (12mA max), whilst the logic required by driver is 7-16mA, but I do not know enough to know what actually happens when these signals go thru a level shifter, if the shifter output can make up the small extra... but anyway my issue is voltage appearing on the driver logic pins when only connecting GND to them... there after when connect the actual step & dir, the voltage levels are not within logic high, and even the ESP32 dir output voltage has dropped ... 

I also suspect that maybe my circuit set up is too crowded (AC too near to DC, wires from 5v PSU pass thru heatsink of driver 😬, etc).. if you can see my repo now, there is a photo of the setup.. I am somewhat restricted in getting all that I need into small (40x30cm) backboard that goes into a electrical box, but was thinking of maybe taking out completely the 80v PSU and installing permanently on top of the box, or one side, but don't like to have exposed connections where people can stick their fingers 🤦

Anyways, I am going to teardown the circuit and test it bit by bit completely separate, then move stuff closer together, if I can get it to work in the first place!

**Reply by gin66** [reply]
Repo-access works now and thanks for sharing. Already stumbled about one piece of information given there. You use pulse counter for encoder. If you use a vanilla  esp32 supporting up to 14 steppers, the first six are using mcpwm and pulse counter. So this may not work at all. So please use the rmt driver.

**Reply by andymid3D** [reply]
Hi gin66, 

as mentioned on the previous discussion about maxSpeed, I will use the line:

stepper = engine.stepperConnectToPin(stepPinStepper, 1); // want to use RMT as want to use independent encoder using PCNT and your docs say RMT is better in the case of choice...

so as not to use the pulse counter with FastAccel... 

I still think this is an issue with current cross over from another component, as when now start to separate the driver as much as the actual wires allow from the 5v psu, now I can get some movement, in one direction, and jittery as hell, when before there was nothing at all.. serial printer also giving strange states of buttons, although the pot is read always just fine...

**Reply by gin66** [reply]
sensitive is most likely the path from esp32 to llc. Perhaps a small capacitor at the input of the llc is helpful.

**Comment by gin66** [comment]
The data sheet of DM860H explains, that the PUL/DIR/ENA are connected to the LED side of the opto couplers. So it expects an input current strong enough to light the LEDs. Either the + inputs are connected to >=5V and the uC pulls (drains) the - inputs to ground, or the - inputs are connected to ground and the uC need to supply sufficient current at e.g. 5V to the + inputs. Judging by the txs0108e data sheet, I doubt the TXS 0108E is a proper part for this job. But I only glanced over the data sheet, so I may be wrong.

**Reply by gin66** [reply]
for any solution the polarity of the step pin should be double checked. The uC generates a low to high transition for a step. The high to low transition can be quite noisy, so should not be used as trigger.

**Reply by andymid3D** [reply]
you write above:

"FastAccelStepper knows only Step/Dir/Enable as digital pins with some flexibility for polarity... "

can I get FAS to use inverted logic signals then.? so I can try supplying 5v to the + inputs..? maybe get a better result there as can wire direct to the 5v PSU +ve terminal, instead of GND, and the uC pulling to GND sounds like a more likely proposition (+ve logic signals were already starting out at 3.3v, plus whatever else was happening adjusting the voltage, but I think when Dir was sent negative, voltage on the Dir driver input was almost 0...🤔

obviously still want to get rid of any crosstalk or EMI, but at the moment even getting the thing to run in both directions would be very encouraging for me!


about your second reply, have no idea how to check polarity of the step pin, AFAIK, it should be +ve as is the dir pin, but I do know that the dir pin is a solid logic 1 or 0 (5v or 0v), whilst with a multimeter, the step pin is going to only give me an average voltage of the pulses being sent as they are much too fast for a multimeter.. I do not have an oscilloscope, nor understand about not using the high to low transition as a trigger...: the DM860 manual on page 5, part 3. Pin assignment, says "each rising or falling edge active", but I presume that it would depend on using +ve (rising) or -ve (falling) inputs for pulse signal..? if I were to try using common anode P- would be a high to low transition, so not good, too noisy...?

**Reply by gin66** [reply]
The manual, which I have found, states: PUL uses rising edge.
https://www.sys-motor.com/Account/Plug-ins/kindeditor/attached/file/20181126/20181126093951_9546.pdf

With the RMT driver inverted logic for step should be possible to implement. For mcpwm/pcnt too complicated. As the mcpwm/pcnt is due to espidf 5.x for me deprecated, that would not be a problem. Just that it is another feature I can not test automatically…

For sure it will be faster to adapt the current logic. This would mean, voltage of >4.5V at PUL while digital output of step pin is high.

**Reply by andymid3D** [reply, marked as answer]
FIXED! although solution is not the expected!
I tried reversing anyway polarity, using common anode, got same results, best case was only one direction  (the OTHER one!) responded, and very jittery... worst case no response
so a friend with more experience in electronics suggested taking the working Nano, and put a pot between signal lines, first Dir to see at what voltage the motor no longer changed direction when pressing the "backwards" button, and then the Pulse signal with an osci to see when the driver stopped responding and the motor no longer spins... the results where surprising:- 1.3v is the driver threshold for HIGH/LOW logic on both Dir and Pulse, so I don't even need the LLC 🤦
Connected directly to ESP32 to driver and now sketch runs exactly the same as on Nano, smooth and stable 👏👏👏
FWIW, with an osci that my friend lent me, also found that the WS2812B Neopixels (5v power) data line signal is 2.2v, so also is working without the supposedly required Level shifter 👏👏👏

Also FWIW, the TXS0108E LLC I continued to try to use for 3 inductive endstops that did absolutely requiere both 5v power and return 5v logic signal, and again started to get strange voltages, specifically on the 5v side of the LLC.. testing w/o anything connected sees 3.3v on all pins on LV side, but either 5v or 1.6-1.7v on HV side, and even measuring the same 5v pin twice in a row have seen this drop.. like it is high then controller chip decides it is floating😳.. with the inductive endstops connected, also was seeing this strange voltage drop on 5v side, but then also the same proportional drop on LV side, so was not getting expected logic levels on pins on ESP32... changed back to a simpler 4 channel mosfet bidirectional level shifter (which did NOT have any effect on the driver issue), and no longer have endstop issues, so have no idea why the TXS0108E LLC behaves the way it does (maybe I managed to fry the onboard chip which decides which pins to put HIGH..? or the driver overloaded something on the 5v side and it went crazy🤷), but have abandoned it as a simpler model LLC gives more reliable results...

I have another new TXS0108E LLC  which has not yet been connected to anything, and later on will be building another machine the same, so may take the opportunity to re-check the situation starting with the inductive endstops on the TXS0108E LLC, just to see if a new one does work as expected, and therefore the previous one I had indeed fried somehow🤷


will also get back when machine is up and running with complete code in my repo, and for sure a video or 2 of it working ;-) 
fingers crossed this will be within this/next week for the basic version (1 mould stored on board, updates to mould struct via Serial from PC), later I will be adding a second ESP32 Display, where will add GUI interaction with the motor uC machine parameters, storage for different mould parameters to be able to send to motor uC, checks on plastic refill amounts/timings, so can control how long each refill has had to heat (and avoid cold injections), and a couple more functions that don't want on the motor-dedicated uC, but still have to learn how to develop  this GUI part, so could take a while

Thanks for the advice so far, and the best Stepper library that have so far tested (for my use case!)

Andy
:-)
